prov/lnx/src/lnx_ep.c

```diff
@@ -0,0 +1,1188 @@
+	struct local_prov_ep *ep;
+	int srq_support = 1;
+	struct lnx_fabric *fabric = lep->le_domain->ld_fabric;
+
+	fi_param_get_bool(&lnx_prov, "use_srq", &srq_support);
+
+	dlist_foreach_container(&fabric->local_prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+			struct local_prov_ep, ep, entry) {
+			if (srq_support) {
+				/* bind the shared receive context */
+				rc = fi_ep_bind(ep->lpe_ep,
+						&ep->lpe_srx.ep_fid.fid, 0);
+				if (rc) {
+					FI_INFO(&lnx_prov, FI_LOG_CORE,
+						"%s doesn't support SRX (%d)\n",
+						ep->lpe_fabric_name, rc);
+					return rc;
+				}
+			}
+
+			rc = fi_enable(ep->lpe_ep);
+			if (rc)
+				return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int lnx_ep_control(struct fid *fid, int command, void *arg)
+{
+	struct lnx_ep *ep;
+	int rc;
+
+	ep = container_of(fid, struct lnx_ep, le_ep.ep_fid.fid);
+
+	switch (command) {
+	case FI_ENABLE:
+		if (ep->le_fclass == FI_CLASS_EP &&
+		    ((ofi_needs_rx(ep->le_ep.caps) && !ep->le_ep.rx_cq) ||
+		    (ofi_needs_tx(ep->le_ep.caps) && !ep->le_ep.tx_cq)))
+			return -FI_ENOCQ;
+		if (!ep->le_peer_tbl)
+			return -FI_ENOAV;
+		rc = lnx_enable_core_eps(ep);
+		break;
+	default:
+		return -FI_ENOSYS;
+	}
+
+	return rc;
+}
+
+int lnx_cq_bind_core_prov(struct fid *fid, struct fid *bfid, uint64_t flags)
+{
+	int rc;
+	struct lnx_ep *lep;
+	/* LNX CQ */
+	struct util_cq *cq;
+	struct local_prov_ep *ep;
+	struct local_prov *entry;
+	struct lnx_fabric *fabric;
+
+	lep = container_of(fid, struct lnx_ep, le_ep.ep_fid.fid);
+	cq = container_of(bfid, struct util_cq, cq_fid.fid);
+	fabric = lep->le_domain->ld_fabric;
+
+	rc = ofi_ep_bind_cq(&lep->le_ep, cq, flags);
+	if (rc)
+		return rc;
+
+	/* bind the core providers to their respective CQs */
+	dlist_foreach_container(&fabric->local_prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+			struct local_prov_ep, ep, entry) {
+			rc = fi_ep_bind(ep->lpe_ep,
+					&ep->lpe_cq.lpc_core_cq->fid, flags);
+			if (rc)
+				return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int lnx_ep_bind_core_prov(struct lnx_fabric *fabric, uint64_t flags)
+{
+	struct local_prov *entry;
+	struct local_prov_ep *ep;
+	int rc;
+
+	dlist_foreach_container(&fabric->local_prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+			struct local_prov_ep, ep, entry) {
+			rc = fi_ep_bind(ep->lpe_ep, &ep->lpe_av->fid, flags);
+			if (rc)
+				return rc;
+		}
+	}
+
+	return rc;
+}
+
+static int
+lnx_ep_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
+{
+	int rc = 0;
+	struct lnx_ep *ep;
+	struct lnx_peer_table *peer_tbl;
+
+	switch (fid->fclass) {
+	case FI_CLASS_EP:	/* Standard EP */
+	case FI_CLASS_SEP:	/* Scalable EP */
+		ep = container_of(fid, struct lnx_ep, le_ep.ep_fid.fid);
+		break;
+
+	default:
+		return -FI_EINVAL;
+	}
+
+	switch (bfid->fclass) {
+	case FI_CLASS_EQ:
+		/* TODO */
+		break;
```

Comment:
This ends up return FI_SUCCESS. This seems incorrect.

Reply:
done